### PR TITLE
env :: support ssh protocol for github repos.

### PR DESCRIPTION
**Pull request by n3f**
Add support for ssh strings in the @wordpress/env package sources. (A similar issue was discussed in #20325).

## What?
Adds URL parsing for ssh strings in `env` `parse-config` module (which returns a `WP_Source` object).

## Why?
Git parsing only supported HTTP (until this PR), but fails to pull/clone authenticated projects.

## How?
The git library already supports ssh; this change just implements parsing into a supported `WP_Source`.

## Testing Instructions
1. Add a ssh url to your `wp-env` enabled project. (e.g. `ssh://git@github.com/some/plugin.git`)
2. Run `<gutenberg-dir>/packages/env/bin/wp-env start`
3. Verify that your plugin was downloaded and enabled in the project.

### CODE REVIEW

File: packages/env/lib/config/parse-config.js
Comment (n3f) [inline packages/env/lib/config/parse-config.js]: Fix

File: packages/env/lib/config/parse-config.js
Comment (noahtallen) [inline packages/env/lib/config/parse-config.js:136]: Does this work for repos that have a different main branch (Like Gutenberg)? Or would you pass, e.g. `ssh://git@github.com/WordPress/WordPress.git#trunk`?

Comment (n3f) [reply packages/env/lib/config/parse-config.js:136]: Yeah, I didn't test and probably something more could be done here. I copied what was done with the HTTP source (i.e. defaulting to master if the ref isn't provided).  Also, if a bad HTTP source is sent it attempts to download and continues into the docker phase (even if a bad ref is provided). (I think, don't quote me on that).  It seemed that users just need to provide a good url. 😆.

A check could be done in the download process if you think that's needed.

Comment (n3f) [reply packages/env/lib/config/parse-config.js:136]: I think the current process works well enough.  Specifying an invalid ref results in:

```sh
$ /gutenberg/packages/env/bin/wp-env start
✖ fatal: couldn't find remote ref nope

GitError: fatal: couldn't find remote ref nope

    at Object.action (/gutenberg/packages/env/node_modules/simple-git/dist/cjs/index.js:1194:25)
    at PluginStore.exec (/gutenberg/packages/env/node_modules/simple-git/dist/cjs/index.js:1229:29)
    at /gutenberg/packages/env/node_modules/simple-git/dist/cjs/index.js:1590:43
    at new Promise (<anonymous>)
    at GitExecutorChain.handleTaskData (/gutenberg/packages/env/node_modules/simple-git/dist/cjs/index.js:1588:16)
    at GitExecutorChain.<anonymous> (/gutenberg/packages/env/node_modules/simple-git/dist/cjs/index.js:1572:44)
    at Generator.next (<anonymous>)
    at fulfilled (/gutenberg/packages/env/node_modules/simple-git/dist/cjs/index.js:55:24) {
  task: {
    commands: [ 'fetch', '--tags', 'origin', 'nope' ],
    format: 'utf-8',
    parser: [Function: parseFetchResult]
  }
}
```

Since this provides a pretty good response and stops the wp-env process it seems that this should be adequate.

**Review by noahtallen — APPROVED** [review]
This tests well for me!

I did notice one little weird thing, but it doesn't cause problems. I think the HTTPS clone is cloning directly to `$wp_env_dir/wordpress`, but the SSH clone went to `$wp_env_dir/wordpress/wordpress`. However, docker-compose still maps it correctly (`$wp_env_path/WordPress/WordPress:/var/www/html`), so it doesn't cause problems.

**Comment by noahtallen** [comment]
Also, would you be able to update the Changelog file in packages/env with this new feature?

**Comment by n3f** [comment]
> Also, would you be able to update the Changelog file in packages/env with this new feature?

I'm not sure about my changelog entry 😆, I put today's date but not sure if that should be different or if everything else looks okay.

I'm not sure how merging occurs, since I don't have the permissions, I assume when it all looks good, you'll just push the button?

File: packages/env/CHANGELOG.md
Comment (noahtallen) [inline packages/env/CHANGELOG.md]:
```suggestion
### Enhancement
-   Added SSH protocol support for git sources
```

I'd mention something about git sources, and we also don't need to add the date (it gets automatically added by the release process, iirc :) )

**Review by noahtallen — APPROVED** [review]
LGTM and still tests well. I think we need to tweak the changelog a bit, but after that I can hit the merge button. (I don't think I have write access to the branch, otherwise I'd update it myself and merge now!)

**Comment by n3f** [comment]
Accepted the suggestion (and added you as a contributor to my fork if anything else needs to be updated).

**Comment by noahtallen** [comment]
Great, thanks! I'll merge once the tests finish :)

**Pull request merged**
